Revamp peg-in transactions

instagibbs opened the pull request:
This PR replaces the overly-complicated pegging methodology with something much simpler.

Old system:
- 21M bitcoin are locked with special templated script code called OP_WITHDRAWPROOFVERIFY
- In order to peg-in the locked outputs are spent in special templated transactions
- The pegged in funds must directly go to the address committed to in the `getpeginaddress` step
- Inputs aren't signed at all. They are simply templated to only allow a single peg-in for the transaction, and no other inputs.
- Peg-in outputs in sidechain must be the exact value of the claimed amount, and the rest must be locked up again.
- Peg-ins cannot pay fees, accordingly.
- Layer violations all over the place.

New system:
- No templating, fewer layer violations, peg-ins can pay fees. Pegins are just "normal" spends of the corresponding Bitcoin output.

Details:
`getpeginaddress` now takes a key in its wallet, creates a witness program, makes SHA2 tweak using that program, then derives the address.

`m_is_pegin` in CTransaction tells the node to check its new `pegin_witness` field that resides alongside the `scriptWitness` field. It expects a certain number of elements on the stack. This includes:

    // 1) value - the value of the pegin output (stored as special 8-byte CScriptNum)
    // 2) asset type - the asset type being pegged in
    // 3) genesis blockhash - genesis block of the parent chain
    // 4) witness program - program that script evaluation will be evaluated against
    // 5) serialized transaction - serialized bitcoin transaction
    // 6) txout proof - merkle proof connecting transaction to header

The Bitcoin peg-in txid/nOut is included as the prevout, which makes sure the txid and signature covers the unique Bitcoin peg-in transaction. We can remove these to save a few bytes since they are completely redundant.

Once the node finds a properly formatted pegin witness, it uses the found witness program, and is directly evaluated as the inputs' scriptpubkey. `SHA256(<witness program>)` is also what the fedredeemscript keys are tweaked by.

The new witness is obviously malleable, which is why `m_is_pegin` is used to cover in signature. Care has to be taken that we aren't banning peers for malleated witnesses or including them into the mempool reject set.

Review action items:
1) Make sure the peg-in database is writing and re-writing which pegins have already occurred correctly.
2) Make sure peg-ins are being considered witness malleable and are treated kindly on the p2p layer
3) Understand the impact on watchman software

Future Work: mininode support

instagibbs commented:
I think I need to re-name "witness_program" and related fields to "witness_scriptpubkey". Program only refers to the hash of the witnessScript.

instagibbs commented:
split up history, should be compiling and passing tests at each commit.

instagibbs commented:
removing WIP tag

arvidn commented on src/coins.cpp:
```diff
@@ -330,6 +330,9 @@ bool CCoinsViewCache::HaveInputs(const CTransaction& tx) const
 {
     if (!tx.IsCoinBase()) {
         for (unsigned int i = 0; i < tx.vin.size(); i++) {
+            if (tx.vin[i].m_is_pegin) {
```
there's no risk with this? essentially to say that we "have" every peg-in input, regardless of if the peg-in is valid or not. (I assume "having" an input normally implies that it's valid/available)

instagibbs replied:
"Have" usually means it exists in the UTXO database, but for peg-in inputs we aren't going to look there. Instead we are going to do two things:
1) Look for a well-formed pegin witness
2) Make sure that peg-in has not already occurred by looking for it in our "spent peg-in" database.

arvidn commented on src/policy/policy.cpp:
```diff
@@ -120,16 +120,11 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
-            }
+        if (tx.vin[i].m_is_pegin) {
+            // This deals with p2sh in general only
             continue;
```
this is probably also something to look out for. non-peg-in inputs are validated to some degree, whereas peg-ins are not validated at all. It probably makes sense to ensure that no caller of this function relies all inputs being validated.

instagibbs replied:
Yes, the trickiest part(versus previous implementation) is making sure that we are handling these cases correctly.

arvidn commented on src/primitives/transaction.h:
```diff
@@ -413,6 +414,7 @@ class CTxIn
     CTxIn()
     {
         nSequence = SEQUENCE_FINAL;
+        m_is_pegin = false;
```
this seems redundant, since you already initialized it at the declaration. (in fact, ``nSequence`` should probably also be initialized at declaration)

arvidn commented on src/primitives/transaction.h:
```diff
@@ -432,10 +434,10 @@ class CTxIn
                 fHasAssetIssuance = false;
```
is it OK that ``prevout.n`` (and later ``output.n``) has the ``COutPoint::OUTPOINT_PEGIN_FLAG`` set here? even if ``m_is_pegin`` is false? Perhaps there's some invariant I'm not aware of, but if ``m_is_pegin`` is always set in this branch of the if, perhaps there should be an ``assert()`` to indicate that

instagibbs replied:
this is a special casing for coinbase inputs. Can you point to the behavior you're worried about exactly? In both directions this case is handled.

arvidn replied:
I get the impression there's an invariant here along the lines of ``m_is_pegin`` <=> ``output.n & COutPoint::OUTPOINT_PEGIN_FLAG``.

But in the (top) if-block (line 431 on the right), this invariant does not seem to be maintained, since ``prevout.n`` can have all bits set, regardless of what ``m_is_pegin`` is. i.e., that branch of the if-statement allows serializing the ``OUTPOINT_PEGIN_FLAG`` regardless.

Perhaps ``m_is_pegin`` is always true if ``prevout.n == -1``. If so, perhaps it would make sense to add an assert to indicate that.

instagibbs replied:
Right, it's another magic value for coinbase input values, which also effects issuance bit logic.

Perhaps more rigorous asserts are needed; any specific places to suggest?

arvidn commented on src/script/interpreter.cpp:
```diff
@@ -1417,182 +1393,7 @@ bool EvalScript(vector<vector<unsigned char> >& stack, const CScript& script, un
-                }
-                break;
-
-                default:
```
+1

instagibbs replied:
what does +1 mean here

arvidn replied:
oh, just that I appreciate this code being removed..

apoelstra replied:
+1 from me too (for this whole file's diff)

instagibbs replied:
my pleasure.

arvidn commented on src/script/standard.h:
```diff
@@ -40,7 +40,7 @@ extern unsigned nMaxDatacarrierBytes;
  * Failing one of these tests may trigger a DoS ban - see CheckInputs() for
  * details.
  */
-static const unsigned int MANDATORY_SCRIPT_VERIFY_FLAGS = SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY | SCRIPT_VERIFY_WITHDRAW;
```
I believe you can remove this constant (``SCRIPT_VERIFY_WITHDRAW``) all together: https://github.com/ElementsProject/elements/blob/elements-0.14.1/src/script/interpreter.h#L113

apoelstra commented on src/validation.cpp:
```diff
@@ -2509,18 +2509,15 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
     if (block.GetHash() == chainparams.GetConsensus().hashGenesisBlock) {
         if (!fJustCheck) {
-            assert(block.vtx.size() == 1);
-            assert(block.nHeight == 0);
```
I think the second assertion can stay.

apoelstra commented on src/validation.cpp:
```diff
@@ -2509,18 +2509,15 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
 
-            CTxUndo undoDummy;
-            UpdateCoins(tx, view, undoDummy, pindex->nHeight);
+            const CTransaction tx = *(block.vtx[block.vtx.size()-1]);
```
Should this be a loop over all coinbase transactions? If we ever have a situation where `AppendPolicyIssuance` is called more than once, this code means that only the last one's outputs are spendable.

apoelstra commented on src/primitives/transaction.cpp:
```diff
@@ -120,6 +121,7 @@ uint256 CTxInWitness::GetHash() const
     leaves.push_back(SerializeHash(vchIssuanceAmountRangeproof, SER_GETHASH, 0));
     leaves.push_back(SerializeHash(vchInflationKeysRangeproof, SER_GETHASH, 0));
     leaves.push_back(SerializeHash(scriptWitness.stack, SER_GETHASH, 0));
+    leaves.push_back(SerializeHash(pegin_witness.stack, SER_GETHASH, 0));
```
Why can't we use the scriptWitness as the pegin_witness?

instagibbs replied:
what's the advantage of doing so? Seems to be abuse, since they serve two separate roles:
`scriptWitness`: utxo spending authorization
`pegin_witness`: peg-in authorization

apoelstra replied:
Avoiding serializing two witness arrays which are never used simultaneously.

instagibbs replied:
But they are. During a peg-in both are used. pegin_witness shows what the "scriptPubKey" is, then scriptWitness satisfies it.

apoelstra replied:
Oh, ok, I hadn't gotten that far yet. Never mind then :)

apoelstra commented on src/validation.cpp:
```diff
@@ -2245,6 +2248,173 @@ bool BitcoindRPCCheck(const bool init)
+bool IsValidPeginWitness(const CScriptWitness& pegin_witness) {
+
+    // Format on stack is as follows:
+    // 1) txid - txid of parent chain transaction - TODO-PEGIN unneeded?
```
I think we should drop the txid, it's cheap to compute and it's an extra thing to validate. Especially now that the transaction is serialized entirely without that 520-byte splitting frustration.

apoelstra replied:
Note that computing the txid does not require understanding Bitcoin transactions. You just have to double-hash the serialized form.

instagibbs replied:
The txid is also already include in the prevout. We should drop it.

instagibbs replied:
as well as the nOut*

apoelstra replied:
Agreed.

apoelstra commented on src/validation.cpp:
```diff
@@ -2245,6 +2248,173 @@ bool BitcoindRPCCheck(const bool init)
+    // Check that the witness program matches the p2ch on the transaction output
+    opcodetype opcodeTmp;
+    unsigned char tweak[32];
+    CSHA256().Write(witness_program.data(), witness_program.size()).Finalize(tweak);
```
The tweak should somehow include the fedpegScript. In the original scheme each pubkey goes into the hash that computes its tweak, which is the standard way to do pay-to-contract and is always secure.

As written, it is possible in principle for the functionaries to adversarially choose their keys to steal some existing Bitcoin output. (They'd be unable to process any other peg-outs, and the existing output would need to be in the correct multisig form, so this is a silly attack scenario, but it is there.)

So there are two options: (1) include `fedpegScript` in its entirety in the tweak, or (2) in the key-tweaking loop recompute the tweak each time using the standard pay-to-contract scheme.

apoelstra commented on src/validation.cpp:
```diff
@@ -1852,23 +1873,39 @@ bool CheckTxInputs(const CTransaction& tx, CValidationState& state, const CCoins
-            assert(coins);
+            if (tx.vin[i].m_is_pegin) {
+                // Check existence and validity of pegin witness
+                if (tx.wit.vtxinwit.size() <= i || !IsValidPeginWitness(tx.wit.vtxinwit[i].pegin_witness) || prevout.hash != uint256(tx.wit.vtxinwit[i].pegin_witness.stack[0]) || (int)prevout.n != CScriptNum(tx.wit.vtxinwit[i].pegin_witness.stack[1], true).getint()) {
```
Won't `prevout.n` have the pegin flag set here?

instagibbs replied:
the flag(as well as the issuance flag) is only set during serialization. During de-serialization it is masked out.

instagibbs commented:
Updated PR addressing the major issues and cleanups.

jonasnick commented on src/validation.cpp:
```diff
@@ -1170,6 +1174,21 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
+                    return state.Invalid(false, REJECT_INVALID, "pegin-no-witness");
+                }
+                std::pair<uint256, COutPoint> pegin = std::make_pair(uint256(tx.wit.vtxinwit[i].pegin_witness.stack[2]), tx.vin[i].prevout);
+                if (view.IsWithdrawSpent(pegin)) {
```
Could a maliciously crafted pegin with `tx.wit.vtxinwit[i].pegin_witness.stack[2].IsNull() == true` crash the node by triggering the assert in `IsWithdrawSpent`?

instagibbs replied:
Yes, for both the prevout txid and the witness item. Great catch. Need to figure out the best way to handle these.

instagibbs commented:
Updated with commit that deals with crash on "null" pegin. Left the asserts in place as they're useful for catching implementation errors.

apoelstra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3388,6 +3389,368 @@ class CSecp256k1Init {
 static CSecp256k1Init instance_of_csecp256k1;
 }
 
+/* Takes federation redeeem script and adds SHA2(witnessProgram) as a tweak to each pubkey */
```
Recommend moving this out of rpcwallet.cpp and making it a general utility function which can also be used in consensus.

Right now this function looks correct but its mirror code (see my comment on https://github.com/ElementsProject/elements/pull/260/commits/65158b7db0445ca541d5e420f75dc0980cb28c62) in peg-in validation is wrong.

BTW is there a test that catches this discrepancy?

instagibbs replied:
the updated code seems to match, you're pointing at an old commit perhaps? Agreed on making it a utility function.

apoelstra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3388,6 +3389,368 @@ class CSecp256k1Init {
+    JSONRPCRequest request2;
+    UniValue varr(UniValue::VARR);
+    varr.push_back(strHex);
+    request2.params = varr;
```
Can we split everything up to here into a `createrawpegin` RPC, since it looks like the standard `sendrawtransaction` and `signrawtransaction` will work to sign and send it? We can leave `claimpegin` which does all three as a convenience method. But as an advanced user it make me nervous when Core produces transactions then signs and publishes them without my review.

apoelstra commented on contrib/assets_tutorial/assets_tutorial.py:
```diff
@@ -91,20 +90,19 @@ def sync_all(e1, e2):
+# Regtest chain starts with 21M bitcoins as OP_TRUE which the wallet
+# understands. This is useful for testing basic functionality and for
+# blockchains that have no pegging functionality. "Policy" coins are required
+# for anti-DoS purposes as well as asset issuance, which consumes inputs for entropy.
```
This sentence needs to be expanded to say what you mean by "policy" coins. At first glance it sounds like it's referring to standardness or something because the paragraph is about output scripts, but then I remembered policyAsset from the source code, which I guess is what you actually mean.

Maybe just replace `"Policy"` with `Initially existing` and the paragraph still works.

instagibbs replied:
No, I specifically meant coins which relay policy use for anti-DoS. I may want to rename the utility functions, since there are other reasons to seed the utxo set.

instagibbs commented:
Rebased, de-duped contract code, made `createrawpegin` RPC call.

maaku commented on src/primitives/transaction.h:
```diff
@@ -261,14 +255,18 @@ class COutPoint
 
+    /* If this flag is set, the CTxIn including this COutPoint
+     * is a peg-in input. */
+    static const uint32_t OUTPOINT_PEGIN_FLAG = (1 << 30);
```
I feel like we've had this discussion in person, and if so I apologize for rehashing it. But why are we reusing an input as a peg-in? Is there some connection between inputs and pegins? Doing it for asset issuance was necessary because we needed a source of entropy. But the outpoint being spent isn't used for the pegin, right?

instagibbs replied:
From a conceptual level, you are "spending" the bitcoin output, after extracting the "real" scriptPubKey from the pegin witness and matching it to the p2ch.

More practically speaking, re-using the prevout fields:
1) Makes each transaction with a peg-in input have a unique txid(assuming there are no double-pegins allowed)
2) Puts those values under the signature

maaku replied:
Ah, because the peg-in data is in the witness. That's fair, although I need to dwell on whether this data should be in the witness or not.

My question was actually more along the lines of "Why not introduce a vector of CTxPegin objects?" But the fact that some pegin data is witness causes this to make sense.

maaku commented on src/primitives/transaction.h:
```diff
@@ -382,6 +380,9 @@ class CTxIn
     CAssetIssuance assetIssuance;
+    /* If this is set to true, the input is interpreted as a
+     * peg-in claim and processed as such */
+    bool m_is_pegin = false;
```
nit: this isn't bitcoin style (`fHas` vs `m_is_`).

instagibbs replied:
style guide has changed somewhat recently: https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md

maaku replied:
Huh. I wonder why. That's an abrupt change from basically ~all of the Satoshi codebase.

maaku commented on src/primitives/transaction.h:
```diff
@@ -501,9 +509,12 @@ class CTxIn
     std::vector<unsigned char> vchInflationKeysRangeproof;
     CScriptWitness scriptWitness;
+    // Re-use script witness struct to include its own witness
+    CScriptWitness pegin_witness;
```
If you're going to follow the bitcoin style guides, this should be `m_pegin_witness`.

maaku commented on src/primitives/transaction.h:
```diff
@@ -513,19 +524,22 @@ class CTxInWitness
     {
-        return vchIssuanceAmountRangeproof.empty() && vchInflationKeysRangeproof.empty() && scriptWitness.IsNull();
+        return vchIssuanceAmountRangeproof.empty() && vchInflationKeysRangeproof.empty() && scriptWitness.IsNull()
+            && pegin_witness.IsNull();
```
nit: either break them all up or make it one line

maaku commented:
Some further notes:

I think that we should hard-fork revamp how we do the transaction inputs, but that it out of scope for this PR and shouldn't hold it up.

I really don't like that this change introduces a consensus rule that pegins can't be repeated. That requires database that grows linearly with the block chain history. This isn't a problem with known existing elements products, but it could become one in the future. A better approach could be to require a log-sized Merkle proof that the pegin outpoint wasn't already claimed, reducing the validator burden to a constant size. But there's enough engineering effort and design required that I could be convinced to move forward on this PR for now, so long as we have agreement on how to address this problem in the near future.

maaku commented on src/validation.cpp:
```diff
@@ -2405,6 +2294,195 @@ bool BitcoindRPCCheck(const bool init)
+    }
+
+    // Get output value. Special 8 byte length to capture possible bitcoin values
+    CAmount value = CScriptNum(stack[0], true, 8).getint64();
```
Why use script serialization here? A VarInt might be better.

jonasnick replied:
Don't we have to catch exceptions from the CScriptNum constructor here?

maaku commented on src/validation.cpp:
```diff
@@ -2405,6 +2294,195 @@ bool BitcoindRPCCheck(const bool init)
+    }
+
+    // Check that the witness program matches the p2ch on the transaction output
+    CScript tweaked_fedpegscript = calculate_contract(Params().GetConsensus().fedpegScript, witness_program);
```
This is assuming that elements script and bitcoin script are the same, or at least similar enough for this to work. That could change in the near future.

maaku commented on src/validation.cpp:
```diff
@@ -2405,6 +2294,195 @@ bool BitcoindRPCCheck(const bool init)
+    CScript scriptDestination(federationRedeemScript);
+    int version;
+    std::vector<unsigned char> program;
+    if (!witnessProgram.IsWitnessProgram(version, program)) {
```
Is this necessary? At some point bitcoin might switch to a new witness commitment format, e.g. when the sixteen remaining version bytes are exhausted or when quantum computers make the security of a 40 byte push untenable. Or for unforeseen reasons.

All you do is feed it into an HMAC later, so why does it care?

maaku commented on src/validation.cpp:
```diff
@@ -2038,14 +1923,25 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
+                    mtx.vout.push_back(GetPeginOutputFromWitness(tx.wit.vtxinwit[i].pegin_witness));
+                    CTransaction tx(mtx);
+                    // Height of "output" in script evaluation will be 0
+                    pegin_coin.FromTx(tx, 0);
```
Doesn't this mean the pegin isn't signing the rest of the transaction?

maaku replied:
Nevermind; I had a brain fart.

instagibbs commented:
> I really don't like that this change introduces a consensus rule that pegins can't be repeated

I don't disagree with the overall point, but we're already doing this once we removed the "reorg proofs". We just need to make sure this isn't a step backwards in making better system.

maaku commented:
> I think I need to re-name "witness_program" and related fields to "witness_scriptpubkey". Program only refers to the hash of the witnessScript.

I would rename this field to `claim_script`. The "witness" part is a complete red herring I think. As mentioned in my review I don't think there's any technical reason to restrict to just witness programs, and we should fix that. But also it doesn't convey what the script's purpose is, which is to authorize a claim.

instagibbs commented:
Rebased onto tip, and I believe have addressed all concerns. I left one test broken(sorry), but it's simply a case of needing to update the exact serialization of the transaction again. Will fix once we're near-final.

jonasnick commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3513,14 +3513,14 @@ UniValue createrawpegin(const JSONRPCRequest& request)
             "1. \"bitcoinTx\"         (string, required) The raw bitcoin transaction (in hex) depositing bitcoin to the mainchain_address generated by getpeginaddress\n"
             "2. \"txoutproof\"        (string, required) A rawtxoutproof (in hex) generated by bitcoind's `gettxoutproof` containing a proof of only bitcoinTx\n"
-            "3. \"witness_program\"   (string, optional) The witness program generated by getpeginaddress. Only needed if not in wallet.\n"
+            "3. \"claim_script\"   (string, optional) The witness program generated by getpeginaddress. Only needed if not in wallet.\n"
```
getpeginaddress generates a "claim script"

jonasnick commented:
Unit tests and pegging.py tests fail at the moment.

A few additional comments:
* I believe that we should require an empty peginwitness if it's not a pegin input. Do we do that?
* There are a few tabs in the PR, for example in IsValidPeginWitness
* DoS rules seem fine to me. Are there specific cases you were thinking of that need more review?
* In general, I have problems understanding CCoinsView but it seems like there is a DoS vector: an attacker could try to claim non-existent Bitcoin transaction resulting in IsWithdrawSpent being called which appears to insert something into a map. We can't DoS ban the attacker so if she repeatedly sends such a transaction she could exhaust resources. This is not a problem specific to this PR though.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3576,7 +3576,7 @@ UniValue createrawpegin(const JSONRPCRequest& request)
         witnessProgScript = CScript(witnessBytes.begin(), witnessBytes.end());
         if (!witnessProgScript.IsWitnessProgram(version, witnessProgram) || version != 0) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "Given witness_program is not a valid v0 witness program.");
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Given claim_script is not a valid v0 witness program.");
```
hmm yes, originally this made sense since it was `claimpegin` functionality which currently assumes this, but it might not make sense for the raw rpc call. Will fix.

instagibbs commented:
tests should be passing now, history is still a bit screwy post-fixes

instagibbs commented:
did history janitor work, will stop messing with it.

instagibbs commented:
Merged.
